src/librbd/api/Snapshot.cc

```diff
@@ -222,6 +226,248 @@ int Snapshot<I>::get_id(I *ictx, const std::string& snap_name, uint64_t *snap_id
     return 0;
   }
 
+template <typename I>
+int Snapshot<I>::list(I *ictx, vector<snap_info_t>& snaps)
+{
+  ldout(ictx->cct, 20) << "snap_list " << ictx << dendl;
+
+  int r = ictx->state->refresh_if_required();
+  if (r < 0)
+    return r;
+
+  std::shared_lock l{ictx->image_lock};
+  for (map<snap_t, SnapInfo>::iterator it = ictx->snap_info.begin();
+       it != ictx->snap_info.end(); ++it) {
+    snap_info_t info;
+    info.name = it->second.name;
+    info.id = it->first;
+    info.size = it->second.size;
+    snaps.push_back(info);
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Snapshot<I>::exists(I *ictx, const cls::rbd::SnapshotNamespace& snap_namespace,
+		        const char *snap_name, bool *exists)
+{
+  ldout(ictx->cct, 20) << "snap_exists " << ictx << " " << snap_name << dendl;
+
+  int r = ictx->state->refresh_if_required();
+  if (r < 0)
+    return r;
+
+  std::shared_lock l{ictx->image_lock};
+  *exists = ictx->get_snap_id(snap_namespace, snap_name) != CEPH_NOSNAP;
+  return 0;
+}
+
+template <typename I>
+int Snapshot<I>::flatten_children(I *ictx, const char* snap_name,
```

Comment:
Nit: I would put this in `librbd/api/Image` since it's a more general image-related function.